File: src/decoders/grade.ts
import { AttachmentKind, type SessionHandle, type Grade } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeGradeValue } from "./grade-value";
import { decodeAttachment } from "./attachment";
import { decodeSubject } from "./subject";

export const decodeGrade = (grade: any, session: SessionHandle): Grade => {
  const id = grade.N;
  const isBonus = grade.estBonus;

  const attachment = (key: string, genre: string) => grade[key] && decodeAttachment({
    G: AttachmentKind.File,
    L: grade[key],
    N: id
  }, session, { G: genre });

  return {
    id,
    value: decodeGradeValue(grade.note.V),
    outOf: decodeGradeValue(grade.bareme.V),
    defaultOutOf: decodeGradeValue(grade.baremeParDefaut.V),
    date: decodePronoteDate(grade.date.V),
    subject: decodeSubject(grade.service.V),
    average: grade.moyenne && decodeGradeValue(grade.moyenne.V),
    max: decodeGradeValue(grade.noteMax.V),
    min: decodeGradeValue(grade.noteMin.V),
    coefficient: grade.coefficient,
    comment: grade.commentaire,
    commentaireSurNote: grade.commentaireSurNote,
    isBonus,
    isOptional: grade.estFacultatif && !isBonus,
    isOutOf20: grade.estRamenerSur20,
    subjectFile: attachment("libelleSujet", "DevoirSujet"),
    correctionFile: attachment("libelleCorrige", "DevoirCorrige")
  };
};

File: library/src/commonMain/kotlin/decoders/grade.kt
package decoders

import kotlinx.serialization.json.*
import models.Attachment
import models.AttachmentKind
import models.Grade
import models.SessionInformation

fun decodeGrade (grade: JsonObject, sessionInformation: SessionInformation): Grade {
    val id = grade["N"]!!.jsonPrimitive.content
    val isBonus = grade["estBonus"]!!.jsonPrimitive.boolean

    val attachment: (key: String, genre: String) -> Attachment? = {key, genre ->
        if (grade[key] == null) null
        else
        decodeAttachment(buildJsonObject {
            put("G", AttachmentKind.File.code)
            put("L", grade[key]!!)
            put("N", id)
        }, sessionInformation, buildJsonObject { put("G", genre) })
    }

    return Grade(
        id = id,
        value = decodeGradeValue(grade["note"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        outOf = decodeGradeValue(grade["bareme"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        defaultOutOf = decodeGradeValue(grade["baremeParDefaut"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        date = decodePronoteDate(grade["date"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        subject = decodeSubject(grade["service"]!!.jsonObject["V"]!!.jsonObject),
        average = if (grade["moyenne"] != null) decodeGradeValue(grade["moyenne"]!!.jsonObject["V"]!!.jsonPrimitive.content) else null,
        max = decodeGradeValue(grade["noteMax"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        min = decodeGradeValue(grade["noteMin"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        coefficient = grade["coefficient"]!!.jsonPrimitive.double,
        comment = grade["commentaire"]!!.jsonPrimitive.content,
        isBonus = isBonus,
        isOptional = (grade["estFacultatif"]?.jsonPrimitive?.boolean ?: false) && !isBonus,
        isOutOf20 = grade["estRamenerSur20"]!!.jsonPrimitive.boolean,
        subjectFile = attachment("libelleSujet", "DevoirSujet"),
        correctionFile = attachment("libelleCorrige", "DevoirCorrige")
    )
}